Add support for writing timestamps without timezone.

sshkvar opened the pull request:
Add `spark.sql.iceberg.store-timestamp-without-zone` spark config to indicate which iceberg type (Types.TimestampType.withZone() or Types.TimestampType.withoutZone()) will be used for spark `TimestampType` type

bkahloon committed:
spark: Add in support to read timestamp without timezone from parquet

bkahloon committed:
spark: Remove ORC vectorized test for reading timestamp without timezone

bkahloon committed:
Spark: address PR comments

bkahloon committed:
Spark: fix build failure due to import of all iceberg packages

bkahloon committed:
Spark: remove unsed imports and try to fix package import ordering

bkahloon committed:
Spark: fix code formatting issue

bkahloon committed:
Spark: fix code formatting issue

bkahloon committed:
Spark: Fix formatting error of long line

bkahloon committed:
Spark: Fix formatting error of long line

sshkvar committed:
Add support for writing timestamps without timezone.
Add `spark.sql.iceberg.store-timestamp-without-zone` spark config to indicate which iceberg type (Types.TimestampType.withZone() or Types.TimestampType.withoutZone()) will be used for spark `TimestampType` type

wangjunyou committed:
Docs: Fix variable name in Flink doc (#2668)

nastra committed:
Nessie: Use AssertJ assertions (#2684)

This also adds AssertJ to testCompile in all modules so assertions can be used elsewhere.

Commit:
Core: add key_metadata to ManifestFile schema and classes (#2675)

Commit:
Docs: Add cache-enabled to catalog property list (#2648)

roryqi committed:
ORC: Remove unused constants in generic readers (#2695)

kbendick committed:
AWS: Fix typo in S3OutputFile.createOrOverwrite exception message (#2699)

rdblue committed:
Update spec for v2 changes (#2654)

* Spec: Add identifier-field-ids to schema.
* Spec: Add section for partition evolution.
* Spec: Add schemas list and current-schema-id to table metadata.
* Spec: Add key_metadata to manifest list.
* Spec: Add schema-id to Snapshot metadata.

Fokko committed:
Parquet: Update to 1.12.0 (#2441)

Parquet changelog: https://github.com/apache/parquet-mr/blob/master/CHANGES.md#version-1120

sshkvar committed:
Added missed check for handling tomestamp without zone for Writer in spark 2.4

sshkvar committed:
Added missed check for handling tomestamp without zone for Writer in spark 2.4

TGooch44 committed:
[python] Adding type ignores for dependencies without type support (#2698)

nastra committed:
Core: Use .as() with AssertJ (#2706)

`.withFailMessage(..)` was mistakenly used and was therefore overriding
the actual error reporting, making debugging difficult.

nastra committed:
API: Add more null checks to TableIdentifier (#2703)

nastra committed:
API: Fix Namespace null handling (#2704)

rominparekh committed:
Docs: Add Adobe Migration article (#2707)

Commit:
Core: Do not allow optional, double, or float identifier fields (#2705)

jun-he committed:
[Python] support custom target name in partition spec builder (#2689)

* support custom target name in partition spec builder

* address the comments.

TGooch44 committed:
Docs: updating README to link to github actions instead of travis-ci.org (#2709)

Co-authored-by: tgooch <tgooch@netflix.com>

nastra committed:
Core: Use equals instead of reference equality (#2714)

See also https://errorprone.info/bugpattern/ReferenceEquality

openinx committed:
Tests: Add unit tests for InternalRecordWrapper, RowDataWrapper, InternalRowWrapper (#2683)

southernriver committed:
Spark: Fix scanAllFiles in MicroBatch.open (#2667)

Param scanAllFiles Used to check whether all the data files should be processed, or only added files.Here we should replace  scanAllFiles to  !scanAllFiles.

coolderli committed:
Docs: Add commit.status-check.* properties (#2661)

Co-authored-by: Ryan Blue <blue@apache.org>

chenjunjiedada committed:
Core: Add delete marker metadata column (#2538)

TGooch44 committed:
[python] Adding Unknown and Void transforms (#2697)

Co-authored-by: tgooch <tgooch@netflix.com>

vinson0526 committed:
fix: add and remove partition transform on same column failed when use v1 metadata (#2691)

RussellSpitzer commented on spark/src/main/java/org/apache/iceberg/spark/SparkUtil.java:
```diff
@@ -20,21 +20,42 @@
+  public static final String HANDLE_TIMESTAMP_WITHOUT_TIMEZONE_SESSION_PROPERTY =
+          "spark.sql.iceberg.handle-timestamp-without-timezone";
+
+  public static final String STORE_TIMESTAMP_WITHOUT_TIMEZONE_SESSION_PROPERTY =
```
I think it's fine to just use one property for both reading and writing

sshkvar replied:
Actually there was incorrect name of this property, changed to `READ_TIMESTAMP_AS_TIMESTAMP_WITHOUT_TIMEZONE`.

This property is responsible for handling how we will represent spark `TimestampType` type in iceberg. By default spark `TimestampType` type will be converted to `Types.TimestampType.withZone()` iceberg type, but if we set READ_TIMESTAMP_AS_TIMESTAMP_WITHOUT_TIMEZONE to true, spark timestamp type will be converted to `Types.TimestampType.withoutZone()`

RussellSpitzer commented on spark/src/main/java/org/apache/iceberg/spark/SparkUtil.java:
```diff
@@ -100,4 +121,56 @@ public static void validatePartitionTransforms(PartitionSpec spec) {
+   */
+  public static boolean hasTimestampWithoutZone(Schema schema) {
+    return TypeUtil.find(schema, t ->
+            t.typeId().equals(Type.TypeID.TIMESTAMP) && !((Types.TimestampType) t).shouldAdjustToUTC()
```
could just check against TimestampType.withoutZone() which returns the singleton instance

sshkvar replied:
good point, changed the code

RussellSpitzer commented on spark/src/main/java/org/apache/iceberg/spark/SparkTypeToType.java:
```diff
@@ -46,12 +46,19 @@
 class SparkTypeToType extends SparkTypeVisitor<Type> {
   private final StructType root;
   private int nextId = 0;
+  private final boolean useTimestampWithoutZone;
```
I'm not sure what this flag is for, I think it's probably safer to always return timestamptype.withZone from here, and handle the mismatch outside of this code

sshkvar replied:
Moved this logic to [SparkFixupTimestampType.java](https://github.com/sshkvar/iceberg/blob/f2fdf2a5f9ffc64b44b353e9fd7d09ed21f8c416/spark/src/main/java/org/apache/iceberg/spark/SparkFixupTimestampType.java)

RussellSpitzer commented on spark/src/main/java/org/apache/iceberg/spark/SparkUtil.java:
```diff
@@ -20,21 +20,42 @@
 
 public class SparkUtil {
+
+  public static final String HANDLE_TIMESTAMP_WITHOUT_TIMEZONE_FLAG = "spark-handle-timestamp-without-timezone";
```
Acutally we could probably just use the same session property here, so we just have one property

spark.sql.iceberg.convert-timestamp-without-timezone

Or something like that

sshkvar replied:
done

RussellSpitzer commented on spark/src/main/java/org/apache/iceberg/spark/TypeToSparkType.java:
```diff
@@ -104,12 +104,14 @@ public DataType primitive(Type.PrimitiveType primitive) {
-        }
-        throw new UnsupportedOperationException(
-            "Spark does not support timestamp without time zone fields");
+//        Types.TimestampType timestamp = (Types.TimestampType) primitive;
```
Need to remove these comment and this is where I think this is where we have the Precondtion.checkArguement(HandleTimeStampWithoutTimezone, Message)

RussellSpitzer commented on spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcWriter.java:
```diff
@@ -118,6 +118,7 @@ public SparkOrcValueWriter primitive(Type.PrimitiveType iPrimitive, TypeDescript
           return SparkOrcValueWriters.decimal(primitive.getPrecision(), primitive.getScale());
         case TIMESTAMP_INSTANT:
+        case TIMESTAMP:
           return SparkOrcValueWriters.timestampTz();
```
Do we need corresponding code in the ParquetWriter as well?

sshkvar replied:
I think no

RussellSpitzer commented on spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -180,6 +184,8 @@ private Expression filterExpression() {
 
   private StructType lazyType() {
     if (type == null) {
+      Preconditions.checkArgument(readTimestampWithoutZone || !SparkUtil.hasTimestampWithoutZone(lazySchema()),
```
Actually cancel what I said about those other spots, this seems like a great place to check for whether we are allowed to do the TZ conversion

sshkvar replied:
ok, so we don't need any changes here?

RussellSpitzer commented on spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java:
```diff
@@ -329,8 +335,10 @@ public boolean enableBatchRead() {
+
       this.readUsingBatch = batchReadsEnabled && hasNoDeleteFiles && (allOrcFileScanTasks ||
-          (allParquetFileScanTasks && atLeastOneColumn && onlyPrimitives));
+              (allParquetFileScanTasks && atLeastOneColumn && onlyPrimitives && !hasTimestampWithoutZone));
```
Why do we need the check here on the batch reading code?

sshkvar replied:
this code was added by @bkahloon 
From my point of view we don't need this check

RussellSpitzer commented on spark3/src/test/java/org/apache/iceberg/spark/sql/TestTimestampWithoutZone.java:
```diff
@@ -0,0 +1,193 @@
+import org.junit.Test;
+import org.junit.runners.Parameterized;
+
+public class TestTimestampWithoutZone extends SparkCatalogTestBase {
```
I think we should be adding this to

```org.apache.iceberg.spark.data.AvroDataTest```

Class as well since it is used in testing data types in TestParquetScan in the Spark Classes

See

https://github.com/apache/iceberg/blob/44c1d00b37f0d7fc2e978baad4f7a861a8335cf0/spark/src/test/java/org/apache/iceberg/spark/data/AvroDataTest.java#L43-L60

I think this will cover ORC and Parquet

sshkvar replied:
added additional test to org.apache.iceberg.spark.data.AvroDataTest

kbendick committed:
[SPARK][BUILD] Allow existing spark2 JMH benchmarks to work with either Spark 2 or Spark 3 (#2595)

Moves spark2 JMH tests to spark directory and update jmh.gradle and build.gradle to allow for running on spark3 and spark2
Makes spark 2 and 3 work with Java 8

nastra committed:
Core: Replace LinkedList usage with ArrayDeque (#2713)

See also https://errorprone.info/bugpattern/JdkObsolete

This also suppresses the same warning for `SchemaUpdate.moveField` as we can't move to an `ArrayDeque` there.

ismailsimsek committed:
Core: Add JDBC catalog implementation (#1870)

kbendick committed:
Core: Throw an error for incremental scans of metadata tables (#2617)

nastra committed:
Core: Use CharSequenceSet instead of Set<CharSequence> (#2712)

This also adds a test that makes sure a CharSequenceSet doesn't suffer
from undefined equality behavior as reported by https://errorprone.info/bugpattern/CollectionUndefinedEquality

Commit:
Spark: Add extensions DDL to set identifier fields (#2560)

kbendick committed:
[AWS] Fix MissingFail error prone warning by cleaning up GlueCatalogCommitFailureTest (#2721)

szehon-ho committed:
Core: Add predicate push down for partitions metadata table (#2358)

Co-authored-by: Szehon Ho <szehon_ho@apple.com>

Commit:
Core: Add HadoopConfigurable interface to serialize custom FileIO (#2678)

Commit:
Docs: Add catalog and metadata files to metadata structure diagram (#2622)

Commit:
AWS: add DynamoDb catalog (#2688)

* AWS: add DynamoDb catalog

* fix spacing, add comment

nastra committed:
Use bulk decryption interface in ArrowReader (#2720)

wypoon committed:
[Spark] Support building against both Spark 3.0 and Spark 3.1. (#2512)

Code changes that allow spark3 and spark3-extensions to be tested against
both Spark 3.0 and Spark 3.1 while still built against a single Spark 3.0 version.

Although additional tests are are created we still only produce a single set of Spark3 binaries which 
are compatible with Spark 3.0 and 3.1

yyanyy committed:
Core: Fix float and double metrics for Parquet and ORC (#2464)

This replaces the metrics from Parquet and ORC with metrics that are accumulated by Iceberg writers to ensure that the metrics do not include NaN values.

nastra committed:
API: Use equals instead of reference equality (#2716)

nastra committed:
Don't use deprecated methods

nastra committed:
Refactor VectorizedArrowReader

nastra committed:
Reduce code duplication in VectorizedColumnIterator

jfz committed:
Spark: Fix check for SQL extensions with extra white space (#2729)

nastra committed:
Reduce code duplication in VectorizedDictionaryEncodedParquetValuesReader

SreeramGarlapati committed:
Spark: Support micro-batch streaming read for DSv2 (#2660)

flyrain committed:
Docs: Remove file spark.md since it is outdated and causes confusion (#2727)

nastra committed:
Reduce code duplication in VectorizedPageIterator

nastra committed:
Reduce code duplication in VectorizedParquetDefinitionLevelReader

stevenzwu committed:
Flink: Rename FlinkTableOptions to more generic FlinkConfigOptions

nastra committed:
Add support for TimeType / UUIDType (#2739)

SreeramGarlapati committed:
Spark: Fix file-open-cost in DSv2 streams (#2743)

yyanyy committed:
Core: Add schema-id to snapshots (#2275)

Commit:
Spec: Fix diagram alignment and size (#2750)

rdblue committed:
Spec: Update for v2 schemas in remaining sections (#2748)

stevenzwu committed:
Flink: Add an optional uidPrefix to FlinkSink#Builder to explicitly set operator uid. (#2745)

sshkvar committed:
Address PR comments.

cwsteinbach committed:
Docs: Update directions for joining #iceberg on Slack (#2758)

RussellSpitzer committed:
Core: Validate planTasks and splitFiles args in TableScanUtil (#2759)


Co-authored-by: Shardul Mahadik <smahadik@linkedin.com>

sshkvar committed:
Address PR comments.

RussellSpitzer committed:
Core: Bin pack strategy cosmetic changes (#2770)

* Core: Refactors to BinPack and Rewrite APIs

Changes to support the implementation of the Spark RewriteDatafileAciton. Fixes JavaDoc links,
adds in BinPackStrategy comments to better binpack real world usecases.

sshkvar committed:
Address PR comments.

sshkvar committed:
Address PR comments.

nastra committed:
Docs: Describe how to configure Code formatter for IntelliJ IDEA (#2766)

samarthjain committed:
Spark: Add limited support for vectorized reads for Parquet V2 (#2749)

With this change, we have added support for Parquet data written in V2 format.
The only data encodings we support are dictionary and plain.
Vectorized reads against data written using Delta/RLE and other encodings are
not supported. As of this commit, note that the Spark Parquet vectorized reads also don't
support vectorized reads for such encodings.

rdblue committed:
Docs: Update for mkdocs 1.2 (#2747)

* Docs: Fix mkdocs use_directory_urls in 1.2.

* Fix broken links and update redirects.

a49a committed:
Docs: Fix typo in flink.md (#2772)

karuppayya committed:
Spark: RemoveReachableFiles action should fail if GC is disabled (#2763)

Co-authored-by: Karuppayya Rajendran <karuppayya.rajendran@apple.com>

nastra committed:
Docs: Describe available Benchmarks and how to run them (#2767)

nastra committed:
Nessie: Properly format code in Nessie module (#2733)

southernriver committed:
Style: Delete blank line  of CachedClientPool.java (#2787)

rdblue committed:
Spec: Update v2 change summary (#2762)

Commit:
Build: bump up DiffPlug Spotless version (#2776)

haormj committed:
Core: Fix JdbcCatalog CATALOG_TABLE_NAME to be lowercase (#2778)

RussellSpitzer committed:
Build: Change Spark Versions to Support M1 Processors (#2795)

Spark's Snappy native lib support is missing M1 support in
our current build. Upgrading Spark upgrades Snappy to a version
which has these native libs. This has no effect on actual runtime
Spark for end users since we do not include Spark with our
release jars.

RussellSpitzer committed:
Docs: Fixes broken links to old spark doc page (#2801)

nastra committed:
Build: Upgrade to JUnit 5 (#2797)

RussellSpitzer committed:
Spark: Reimplement RewriteDatafilesAction with partial progress (#2591)

sshkvar committed:
Address few little clean up

marton-bod committed:
Upgrade to Tez 0.10.1 (#2790)

sshkvar committed:
Address few little clean up

sshkvar committed:
Address few little clean up

sshkvar committed:
fix for `'lambda arguments' has incorrect indentation level 12, expected level should be one of the following: 6, 8. `

sshkvar committed:
fix for `incorrect indentation level 6, expected level should be 8.`

RussellSpitzer commented:
You can run ```g build -x test  -x integrationTest -x testSpark31``` to check the static analysis in the future

Commit:
Spark: Parallelize task init when fetching locality info (#2800)

daksha121 committed:
Spark: Add table property to skip delete snapshots in streaming (#2752)

aokolnychyi committed:
API: Use delete instead of remove in action names (#2810)

aokolnychyi committed:
Spark: Use JavaSparkContext.fromSparkContext instead of constructor (#2812)

aokolnychyi committed:
Spark: Add missing deprecation annotations for old actions (#2811)

sshkvar committed:
Added withSQLConf method to AvroDataTest.java as suggested in the PR comments.
Fixed code formatting

nastra committed:
Docs: Fix link to intellij-java-palantir-style.xml (#2817)

szehon-ho committed:
Spark : Add Files Perf improvement by push down partition filter to Spark/Hive catalog (#2777)

Pushes down partition filters in Spark/Hive Import to underlying catalog instead of retrieving all partitions and then filtering.

Fokko committed:
Core: Use Avro 1.10.1 (#1648)

Co-authored-by: Fokko Driesprong <fdriesprong@ebay.com>

RussellSpitzer commented:
@sshkvar It seems like this PR is out of sync with master so I can't merge it, can you rebase it?

sshkvar commented:
@RussellSpitzer looks that something went wrong with rebase, 320 files changes which is incorrect

sshkvar commented:
> @sshkvar It seems like this PR is out of sync with master so I can't merge it, can you rebase it?

I thinks I need to close this PR, main PR is https://github.com/apache/iceberg/pull/2757
